File: api/functions/poi.js
const puppeteer = require("puppeteer");
let obj = require("../data/object");

module.exports = async function (location) {
  try {
    const browser = await puppeteer.launch({
      executablePath: "/usr/bin/chromium-browser",
    });
    const page = await browser.newPage();
    await page.goto(
      `https://www.holidify.com/places/${location}/sightseeing-and-things-to-do.html`
    );
    let data = await page.evaluate(() => {
      let results = [];
      let items = document.querySelectorAll(".content-card");
      items.forEach((items) => {
        results.push({
          foto_url: items
            .querySelector(".card-img-top")
            .getAttribute("data-original"),
          title: items.querySelector(".card-heading").innerHTML,
        });
      });
      return results;
    });
    obj["Point_of_Interests"]=[];
    if (data.length > 0) obj["Point_of_Interests"] = data;
    // console.log(data);
    await browser.close();
  } catch (error) {
    console.error(error);
  }
};

File: travelapp/lib/api/poi.dart
import 'package:html/parser.dart' as parser;
import 'package:http/http.dart' as http;

import '../models/location_poi.dart';

Future<List<PointOfInterest>> extractPOIData(String? location) async {
  // Getting the response from the targeted url
  List<PointOfInterest> listPOI = [];
  final response = await http.Client().get(Uri.parse(
      "https://www.holidify.com/places/$location/sightseeing-and-things-to-do.html"));

  // Status Code 200 means response has been received successfully
  if (response.statusCode == 200) {
    // Getting the html document from the response
    var document = parser.parse(response.body);
    try {
      // Scraping the first article title
      var response = document.getElementsByClassName('content-card');

      for (int i = 0; i < response.length; i++) {
        var title =
            response[i].getElementsByClassName("card-heading")[0].text.trim();
        var url = response[i]
            .getElementsByClassName("card-img-top")[0]
            .attributes["data-original"];
        var desc =
            response[i].getElementsByClassName("card-text")[0].text.trim();
        listPOI.add(PointOfInterest(fotoUrl: url, title: title, desc: desc));
      }
      return listPOI;
    } catch (e) {
      return [];
    }
  } else {
    return [];
  }
}
